### PR TITLE
Add some basic testground tests

### PR DESCRIPTION
**Pull request by merlinran**
Here are the tasks breakdown:
- [x]  Have a bunch of nodes each creating tons of records in their own threads, then make sure all nodes have the same view eventually. 1d
- [x]  Same as the above, but have the records referencing each other nodes' records in different threads. 1d
- [x]  Given an already fully synced network, make sure a newly joined node can be fully synced with the rest. 0.5d
- [x]  Record the time metrics of each test and store them as the baseline for further improvements. 1d
- [x]  Add variations of the plans to introduce slow / disconnected / failing intermittently nodes. 0.5d

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -35,19 +35,15 @@ require (
 	github.com/ipfs/go-ipld-cbor v0.0.5
 	github.com/ipfs/go-ipld-format v0.2.0
 	github.com/ipfs/go-log v1.0.4
-	github.com/ipfs/go-log/v2 v2.1.1
+	github.com/ipfs/go-log/v2 v2.1.3
 	github.com/ipfs/go-merkledag v0.3.2
-	github.com/libp2p/go-libp2p v0.12.0
+	github.com/libp2p/go-libp2p v0.14.0
```
Comment (merlinran) [inline go.mod:40]: Roman told me this v0.14.0 solved many simultaneous connections issues for Anytype, so I tried here. Didn't improve significantly, but not bad to have I suppose?

File: integrationtests/testground/README.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+#  This one is good for local test while developing.
+testground run composition --wait -f current-exec.toml
+```
+
+You can also run individual test case and pass test parameters in command line. For example, below builds the test as native executables, runs 2 instances, with very verbose logs printed to the console. See manifest.toml for all test parameters.
+```
+testground run single --wait -p go-threads -t sync-threads -b exec:go -r local:exec -i 2 -tp verbose=2
+```
+
+# Analysize
```
Comment (merlinran) [inline integrationtests/testground/README.md:38]: Still need to figure out how to access the Grafana and InfluxDB containers started along with testground daemon, but that's most useful in hosted environment, so we can skip atm.

File: integrationtests/testground/main.go [hunk 3]
```diff
@@ -0,0 +1,453 @@
+// This is the test program to be run by testground. Here's what it does:
```
Comment (merlinran) [inline integrationtests/testground/main.go:1]: The code logic is still a bit obscure, but that is the result of several iterations trying to make it more understandable, so I add this overview to help.

File: integrationtests/testground/main.go [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+	}
+	fail = func(msg string, args ...interface{}) {
+		env.RecordFailure(fmt.Errorf(msg, args...))
+	}
+	setup(env, ic)
+	for i, pair := range simulatedNetworks {
+		round := pair.simulation
+		if ic.GlobalSeq == 1 {
+			msg("################### Round %s network ###################", round)
+			ic.NetClient.MustConfigureNetwork(context.Background(), &network.Config{
```
Comment (merlinran) [inline integrationtests/testground/main.go:114]: See https://docs.testground.ai/traffic-shaping#configure-traffic-shaping

File: integrationtests/testground/main.go [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+		return
+	}
+	myRecords := env.IntParam("records")
+	meLive := 1
+	if env.BooleanParam("late-start") {
+		myRecords = 0
+		meLive = 0
+	}
+	expectedRecords := publishAndGetTotal(sync.NewTopic("expectedRecords-"+round+"-step-1", 0), myRecords)
+	livePeers := publishAndGetTotal(sync.NewTopic("livePeers-"+round+"-step-1", 0), meLive)
```
Comment (merlinran) [inline integrationtests/testground/main.go:188]: Each instance runs independently, and there's no way for an instance to get the full picture of how many instances in each group, so the instances need to communicate with each other to get the correct number.

File: integrationtests/testground/main.go [hunk 6]
```diff
@@ -0,0 +1,453 @@
...
+			msg("Got all %d records", nRecords)
+			return
+		}
+	}
+	return
+}
+
+func (t *threadWithKeys) AddRecord(ctx context.Context, prev cid.Cid) (rec corenet.Record, err error) {
+	obj := make(map[string][]byte)
+	fuzzer.Fuzz(&obj)
```
Comment (merlinran) [inline integrationtests/testground/main.go:426]: This is to create payload of different sizes / edge cases. Not meant to fuzz test cbor implementation. Just want to add some variance to the network traffic.

File: net/api/client/client_test.go [hunk 7]
```diff
@@ -411,58 +408,6 @@ func setup(t *testing.T) (ma.Multiaddr, *Client, func()) {
 	}
 }
 
-func makeServer(t *testing.T) (ma.Multiaddr, ma.Multiaddr, func()) {
```
Comment (merlinran) [inline net/api/client/client_test.go:414]: Moved to test_util.go as `CreateTestService` for reuse.

File: net/pubsub.go [hunk 8]
```diff
@@ -168,6 +170,9 @@ func (s *PubSub) subscribe(ctx context.Context, id thread.ID, topic *topic) {
 		}
 		log.Debugf("received multicast record from %s", from)
 
+		ctx = grpcpeer.NewContext(ctx, &grpcpeer.Peer{
```
Comment (merlinran) [inline net/pubsub.go:173]: This is a fix by Sander.

**Review by merlinran — COMMENTED** [review]
Okay this is finally ready for review, but the test is very very unreliable. The only way to run it with a high chance of success is via below. Two peers each with 2 initial records!
```
testground run single --wait -p go-threads -t sync-threads -b exec:go -r local:exec -i 2 -tp records=2
```

The recommended way of running test locally while developing is
```
testground run composition --wait -f current-exec.toml
```
But I never made it run through a full set of network simulations, even though those "simulations" have no effect in `local:exec` mode.

Anyway I think the test code is somewhat solid. And with this we can tweak the code to pass test more reliably.

**Comment by sanderpick** [comment]
Nice! Will look this evening.

**Comment by sanderpick** [comment]
👀

File: integrationtests/testground/main.go (hunk 5 above)
Comment (sanderpick) [reply integrationtests/testground/main.go:188]: So this is some kind of pubsub between instances?

File: integrationtests/testground/main.go [hunk 9]
```diff
@@ -0,0 +1,451 @@
...
+import (
+	"context"
+	crand "crypto/rand"
+	"fmt"
+	"net"
+	"net/http"
+	_ "net/http/pprof"
+	"time"
+
+	fuzz "github.com/google/gofuzz"
```
Comment (sanderpick) [inline integrationtests/testground/main.go:24]: Nice looking lib

File: integrationtests/testground/main.go [hunk 10]
```diff
@@ -0,0 +1,451 @@
...
+				CallbackState: sync.State("network-configured-" + round),
+				RoutingPolicy: network.AllowAll,
+			})
+			msg("Done configuring %s network", round)
+		}
+
+		desiredAddr := ""
+		if env.TestSidecar {
+			// listen on non-local interface when running in container, so the peers can communicate with each other
+			ip := ic.NetClient.MustGetDataNetworkIP().String()
```
Comment (sanderpick) [inline integrationtests/testground/main.go:127]: Slick

**Review by sanderpick — APPROVED** [review]
Very nice! We've been dreaming of TG tests for awhile... really excited to see this landing :)

**Comment by sanderpick** [comment]
I re-ran the tests. The failing test from previous run looks like an intermittent thing that I thought was resolved in the past:
```
--- FAIL: TestModifiedSince (0.00s)
    --- FAIL: TestModifiedSince/WithCreateSaveAndDelete (1.19s)
Error:         collection_test.go:956: should have modfied id 01f65p2j6x7d6phffkzdbb2asd
```
cc @carsonfarmer if that rings any bells.

File: integrationtests/testground/main.go (hunk 5 above)
Comment (merlinran) [reply integrationtests/testground/main.go:188]: It's more like `atomic.AddInt64()` across processes, implemented by Testground via a central Redis.

**Comment by merlinran** [comment]
Just realized this was not merged. The failed test is not new, but will be taken care of as part of this sprint.

**Comment by sanderpick** [comment]
Oh cool. I just created an issue for that failure. https://github.com/textileio/go-threads/issues/520

**Comment by carsonfarmer** [comment]
I think I have a really simple fix for this. @merlinran do you want me to PR the fix here? Alternatively, I can just past the fix below and you can sneak it in yourself?

**Comment by merlinran** [comment]
Either way works @carsonfarmer . If you paste here, I'll add it to my ongoing PR to improve testground tests reliability.

**Comment by carsonfarmer** [comment]
Cool, line `955` in `db/collection_test.go` (TestModifiedSince):

```
		if mods[0] != res[0] && mods[1] != res[0] {
			t.Fatalf("should have modified id %s", res[0])
		}
```

The ordering is arbitrary. The test was overly naive, that should catch either one being the correct value,

**Comment by merlinran** [comment]
Done https://github.com/textileio/go-threads/commit/17873f261d501a135b33b898e0c9d01a1591f104
